systemtest/benchtest/expvarmetrics/metrics.go

```diff
@@ -0,0 +1,253 @@
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package expvarmetrics
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+)
+
+type Metric int
+
+const (
+	Goroutines Metric = iota
+	Bytes
+	MemAllocs
+	MemBytes
+	RSSMemoryBytes
+	HeapAlloc
+	HeapObjects
+	NumGC
+	PauseTotalNs
+	AvailableBulkRequests
+	TotalEvents
+	ActiveEvents
+	TransactionsProcessed
+	SpansProcessed
+	MetricsProcessed
+	ErrorsProcessed
+	ErrorElasticResponses
+	ErrorOTLPTracesResponses
+	ErrorOTLPMetricsResponses
+)
+
+type AggregateStats struct {
+	First       int64
+	Last        int64
+	Min         int64
+	Max         int64
+	Mean        float64
+	sampleCount int64
+}
+
+type watchItem struct {
+	notifyChan chan<- bool
+	validator  func(int64) bool
+}
+
+// Collector defines a metric collector which queries expvar
+// endpoints and aggregates the collected metrics.
+//
+// Watchers are one-time hooks into collector to know about when
+// the a specific state of the metric is observed by the collector.
+type Collector struct {
+	l         sync.Mutex
+	serverURL string
+	period    time.Duration
+	metrics   map[Metric]*AggregateStats
+	watchers  map[Metric]*watchItem
+}
+
+// StartNewCollector creates a new collector and starts
+// querying expvar endpoint at the specified interval.
+func StartNewCollector(ctx context.Context, serverURL string, period time.Duration) (*Collector, error) {
+	c := &Collector{
+		serverURL: serverURL,
+		period:    period,
+		metrics:   make(map[Metric]*AggregateStats),
+		watchers:  make(map[Metric]*watchItem),
+	}
+	return c, c.start(ctx)
+}
+
+// Get returns the aggregated stat of a given metric.
+func (c *Collector) Get(m Metric) AggregateStats {
+	c.l.Lock()
+	defer c.l.Unlock()
+
+	stats := c.metrics[m]
+	statsCopy := *stats
+	return statsCopy
+}
+
+// Delta returns the diff of a metric valuefrom the first
+// event that was collected by the collector.
+func (c *Collector) Delta(m Metric) int64 {
+	c.l.Lock()
+	defer c.l.Unlock()
+
+	stats := c.metrics[m]
+	return stats.Last - stats.First
+}
+
+// AddWatch configures a new watcher for a given metric. The
+// watcher is deleted after the specified expected value is
+// observed.
+//
+// Watcher returns a read only channel to observe the state
+// of the watch. A true event on this channel refers to an
+// observation with the expected value while a false event
+// refers to an error in the collector. Both cases end the
+// lifecycle of a watcher.
+func (c *Collector) AddWatch(m Metric, expected int64) (<-chan bool, error) {
+	c.l.Lock()
+	defer c.l.Unlock()
+
+	if _, ok := c.watchers[m]; ok {
+		return nil, errors.New("watcher already exists for the given metric")
+	}
+
+	out := make(chan bool)
```

Comment:
We should change the channel to be buffered, in case the condition is reached, but there aren't any receivers actively listening for the change, otherwise, when the condition is reached, it'll block forever on the `watcher.notifyChan <- true` statement.

```suggestion
	// Create a buffered channel in case there aren't any receivers when the
	// condition is reached.
	out := make(chan bool, 1)
```